File: 2024/day01.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <chrono>

int main() {
    // start clock
    auto start = std::chrono::high_resolution_clock::now();

    int part01 = 0;
    int part02 = 0;

    std::vector<int> leftList;
    std::vector<int> rightList;

    // file handling
    std::string line;
    std::ifstream input;
    input.open("input/day01.txt");
    while (getline(input, line)) {
        // seperate left and right columns to different vectors
        leftList.push_back(std::stoi(line.substr(0, line.find("   "))));
        rightList.push_back(std::stoi(line.substr(line.find("   "))));
    }
    input.close();

    // sort vectors
    std::sort(leftList.begin(), leftList.end());
    std::sort(rightList.begin(), rightList.end());

    // solve for part 1 and 2
    for (int i = 0; i < leftList.size(); i++) {
        part01 += abs(leftList[i] - rightList[i]);
        // counts instances of leftList[i] in the rightList vector
        part02 += leftList[i] * std::count(rightList.begin(), rightList.end(), leftList[i]);
    }

    // stop clock
    auto stop = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start).count();

    std::cout << "part 1: " << part01 << "\n";
    std::cout << "part 2: " << part02 << "\n";
    std::cout << "time taken: " << duration << "ms\n";
    return 0;
}

File: 2025/day01.py
from time import time

def solve():
    with open("input/day01.txt", "r") as f:
        data = [line.strip() for line in f.readlines()]
    f.close()

    part01 = 0
    part02 = 0
    
    current_number = 50
    for instruction in data:
        direction = 1 if instruction[0] == "R" else -1
        length = int(instruction[1:len(instruction)])

        current_number += direction * length

        if current_number > 100:
            part02 += int(current_number / 100)
            if current_number - length == 100 or current_number % 100 == 0:
                part02 -= 1
        elif current_number < 0:
            part02 += int(abs(current_number - 100) / 100)
            if current_number + length == 0 or current_number % 100 == 0:
                part02 -= 1

        current_number %= 100

        if (current_number == 0):
            part01 += 1
            part02 += 1

    return part01, part02

start_time = time()
part01, part02 = solve()
stop_time = time()

print("part 1:\t\t", part01)
print("part 2:\t\t", part02)
print("duration:\t", round(1000000 * (stop_time - start_time)), "µs")
